sys/net/application_layer/nanocoap/sock.c

```diff
@@ -131,6 +132,111 @@ ssize_t nanocoap_get(sock_udp_t *sock, const char *path, uint8_t *buf, size_t le
     return res;
 }
 
+static int _fetch_block(coap_pkt_t *pkt, uint8_t *buf, sock_udp_t *sock,
+                        const char *path, coap_blksize_t blksize, size_t num)
+{
+    uint8_t *pktpos = buf;
+    uint16_t lastonum = 0;
+
+    pkt->hdr = (coap_hdr_t *)buf;
+
+    pktpos += coap_build_hdr(pkt->hdr, COAP_TYPE_CON, NULL, 0, COAP_METHOD_GET,
+                             num);
+    pktpos += coap_opt_put_uri_pathquery(pktpos, &lastonum, path);
+    pktpos += coap_opt_put_uint(pktpos, lastonum, COAP_OPT_BLOCK2,
+                                (num << 4) | blksize);
+
+    pkt->payload = pktpos;
+    pkt->payload_len = 0;
+
+    int res = nanocoap_request(sock, pkt, NANOCOAP_BLOCKWISE_BUF(blksize));
+    if (res < 0) {
+        return res;
+    }
+
+    res = coap_get_code(pkt);
```

Comment:
This is by far insufficient to determine the success of the request; at minimum it'd need to look at response options and see if any are critical. Moreover, if an ETag is present (which the server SHOULD provide), it MUST be compared as per <https://datatracker.ietf.org/doc/html/rfc7959#section-2.4>.

(I know this is only being moved around, but one might manage to convince oneself that SUIT servers always produce immutable resources and/or such that are checked by the client anyway, but now it's becoming a generic tool).

Reply:
Since we are not doing *any* option processing here, I assume it is enough to send a 4.02 (Bad Option) response on *any* critical option here?

Reply:
Ah there is of course the Block2 option that we consume.
Now we might do something like

```c
static bool _has_crit_opt(const coap_pkt_t *pkt)
{
    /* we don't process any options, so just check if there is a critical one */
    for (unsigned i = 0; i < pkt->options_len; ++i) {
        /* ignore non-critical options */
        if ((pkt->options[i].opt_num & 1) == 0) {
            continue;
        }

        switch (pkt->options[i].opt_num) {
        case COAP_OPT_BLOCK2:
            continue;
        default:
            return true;
        }
    }

    return false;
}
```

But that feels pretty hacky.
Or we do

```C
/**
 * @brief   CoAP PDU parsing context structure
 */
typedef struct {
    coap_hdr_t *hdr;                                  /**< pointer to raw packet   */
    uint8_t *token;                                   /**< pointer to token        */
    uint8_t *payload;                                 /**< pointer to payload      */
    uint16_t payload_len;                             /**< length of payload       */
    uint16_t options_len;                             /**< length of options array */
    coap_optpos_t options[CONFIG_NANOCOAP_NOPTS_MAX]; /**< option offset array     */
    uint64_t opt_consumed;                            /**< bit field               */
#ifdef MODULE_GCOAP
    uint32_t observe_value;                           /**< observe value           */
#endif
} coap_pkt_t;
```

to mark which options have been consumed and then check if there are any critical options left unconsumed.

But is this still in the spirit of nanocoap?

Reply:
> But that feels pretty hacky.

That's to some extent because no CoAP thing in RIOT does this right yet.

> But is this still in the spirit of nanocoap?

That I don't feel like I'm the right person to answer. I've thought occasionally about the bitfields but always thought them inconveniently indexed by option numbers (ranging to 64k so not practical), indexing them by option position is convenient as they're bound in number. A hybrid solution could set only the critical ones at parsing time, they're ticked off at processing (maybe even by accessing) and then there's a check for done by whoever says that now everything that'll be processed has been.

Reply:
I've added the check f85102c3f7911dd3ca6685f8f07ff139cf53b80f
I tried to keep the run-time and memory costs at a minimum.

Reply:
That LGTM implementation-wise, but the API change of the const being removed will need justification in documentation, and `_has_crit_opt` should probably become a public `coap_has_unprocessed_critical_options`.

Given @kaspar030 is, like, the guardian of nanocoap not blowing up, I'd like to summon him for his blessing.

I can offer documentation for that:

> Check whether any of the packet's options that are critical (i.e. must be understood by the receiver, indicated by a 1 in the option number's least significant bit) were not accessed since the packet was parsed.
>
> Call this in a server on requests after all their option processing has happened, and stop processing the request if it returns `true`, returning a 4.02 Bad Option response.
>
> Call this in a client when receiving a response before acting on it; consider the response unprocessable if it returns `true`.
>
> @return        `true` if any of the options marked as critical at parse time have not been accessed.
> @return        `false` if there are no critical options, or all have been accessed.

This should be accompanied by a general note in the nanocoap API:

> Packets accessed through @ref coap_find_option or any of the `coap_opt_get_*` functions track their access in bit field created at parsing time to enable checking for critical options in @ref coap_has_unprocessed_critical_options. These functions thus have a side effect, and code that calls them on critical options needs to ensure that failure to process the accessed option is propagated into failure to process the message. For example, a server helper that tries to read the If-None-Match option (which is critical) and finds it to be longer than it can process must not return as if no If-None-Match option was present, as it has already triggered the side effect of marking the option as processed.

(With apologies to the poor GitHub users who were so unfortunate as to pick user names that coincide with Doxygen commands: let me know where to sign the petition for better plain-text markup formats.)

Reply:
Thank you, added